src/Mvc/Mvc.Core/src/Formatters/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,166 @@
+        {
+            ThrowArgumentException(buffer, offset, count);
+            var bufferSegment = new ArraySegment<byte>(buffer, offset, count);
+            return WriteAsync(bufferSegment, flush: false, cancellationToken);
+        }
+
+        private async Task WriteAsync(
+            ArraySegment<byte> bufferSegment,
+            bool flush,
+            CancellationToken cancellationToken)
+        {
+            var decoderCompleted = false;
+            while (!decoderCompleted)
+            {
+                _decoder.Convert(
+                    bufferSegment,
+                    _charBuffer.AsSpan(_charsDecoded),
+                    flush,
+                    out var bytesDecoded,
+                    out var charsDecoded,
+                    out decoderCompleted);
+
+                _charsDecoded += charsDecoded;
+                bufferSegment = bufferSegment.Slice(bytesDecoded);
+
+                if (flush || !decoderCompleted)
+                {
+                    // This is being invoked from FlushAsync or the char buffer is not large enough
+                    // to accomodate all writes.
+                    await WriteBufferAsync(flush, cancellationToken);
+                }
+            }
+        }
+
+        private async Task WriteBufferAsync(bool flush, CancellationToken cancellationToken)
+        {
+            var encoderCompletd = false;
+            var charsWritten = 0;
+            var byteBuffer = ArrayPool<byte>.Shared.Rent(_maxByteBufferSize);
+
+            try
+            {
+                while (!encoderCompletd && charsWritten < _charsDecoded)
+                {
+                    _encoder.Convert(
+                        _charBuffer.AsSpan(charsWritten, _charsDecoded - charsWritten),
+                        byteBuffer,
+                        flush,
+                        out var charsEncoded,
+                        out var bytesUsed,
+                        out encoderCompletd);
+
+                    await _stream.WriteAsync(byteBuffer.AsMemory(0, bytesUsed), cancellationToken);
+                    charsWritten += charsEncoded;
+                }
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(byteBuffer);
+            }
+
+            // At this point, we've written all the buffered chars to the underlying Stream.
+            _charsDecoded = 0;
+        }
+
+        private static void ThrowArgumentException(byte[] buffer, int offset, int count)
+        {
+            if (count <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (offset <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
```

Comment:
nit: Why did you decide on throwing with count? What if `offset > buffer.Length` and `count = 0`?